Implementation Plan: Copy updates `mature` -> `sensitive` (#2126)
WordPress/openverse: documentation/projects/proposals/trust_and_safety/content_report_moderation/20230517-implementation_plan_mature_copy_change.md

to copy/text itself. Field and model names for the API should also be updated
when possible, but without the need for database migrations (i.e., using
[Django's built-in tools for specifying different underlying names for fields and tables](https://docs.djangoproject.com/en/4.2/ref/models/options/#table-names)).

## Expected Outcomes

<!-- List any succinct expected products from this implementation plan. -->

All references to `mature` in code and copy will be updated to `sensitive`,
except where `mature` is
[part of a data model](https://github.com/WordPress/openverse/blob/0f90f84d4bf856297356a918e4cc4fe42d07ffc2/ingestion_server/ingestion_server/elasticsearch_models.py#L118)
or makes the most sense to keep (e.g. when interacting with certain provider
aspects).

## Outlined Steps

<!-- Describe the implementation step necessary for completion. -->

The work is broken up into two services (API and frontend) and those services
are broken into two aspects (code and copy).

### API

#### Code

The following models will need to be updated. Since we want to avoid a migration
with this work, they will need to

> [explicitly reference the old table names using `Meta::db_table`](https://docs.djangoproject.com/en/4.2/ref/models/options/#table-names).

Reviewer: I'll also support a version of this IP with migrations. No need to do it now but it would be a plus to keep the DB model updated too.

Reviewer 2: It would probably be better to do sooner than later, if it's decided that that is desirable. It'll only get more tedious as the tables grow in size, right? Specifically with respect to the fact that we'd need to keep two tables for the rename to work with zero-downtime and copying the data will take longer as the tables grow. They're all pretty small (relatively speaking) at the moment, however.

Author: Agreed, if we were going to change them now would be the time. I can rework this a bit to include instructions for how to approach that.

- [`AbstractMatureMedia`](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/api/api/models/media.py#L334-L333)
- [`MatureImage`](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/api/api/models/image.py#L83)
- [`MatureAudio`](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/api/api/models/audio.py#L263)

The
[`mature` reason](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/api/api/models/media.py#L22-L21)
(and
[`mature_filtered`](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/api/api/models/media.py#L18-L17)
value) in `AbstractMediaReport` definition will also need to be changed to
`sensitive` and `sensitive_filtered` respectively. In order to remain backwards
compatible with the frontend, incoming references on reports for `mature` will

> need to be mapped to the new value.

Reviewer 2: If I'm understanding the approach suggested here, this would have us update all the previous report values, right? I think that will require multiple steps including Django migrations (Django insists on encoding "choices" into migrations even though Postgres doesn't know anything about it) and (even if Django migrations aren't needed) a management command to update the old values (as per [our documentation on data transformations](https://docs.openverse.org/general/zero_downtime_database_management.html#how-to-achieve-zero-downtime-deployments)). Is my understanding of the suggestion correct? If so I think it would be worth outlining the steps for that explicitly.

FWIW, I don't think this violates the no-migrations stipulation from the project proposal as that was aimed primarily at table and column name changes rather than _data_ changes. Table and column name changes are just a lot harder to juggle and can be masked in code and UI using Django's tools. The data inside the columns, however, is easier to update and much peskier if we don't (easy to accidentally mess up queries, etc).

Author: Oh that's very true, thank you for mentioning it! I'll add that to this IP 🙂

Per the
[detecting sensitive textual content project plan](https://docs.openverse.org/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.html#designation-of-results-with-sensitive-terms),
the `mature` field of the media serializer will remain, and will eventually have
a `sensitivity` field added to it to encapsulate specific information about the
reason behind a media being marked sensitive. Those changes are not under the
purview of this IP.

The Elasticsearch model property
[`mature`](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/ingestion_server/ingestion_server/elasticsearch_models.py#L118)

> will also not be affected.

Reviewer: Will the ES model property also change later in conjunction with the field addition to the serializer?

Author: I'm not sure! I think `sensitive` is actually something we might end up deriving from a few different places in Elasticsearch, and might not be something we store on the indexed records directly. @sarayourfriend did you have a different impression?

Reviewer 2: `mature` is already a derived property. It theoretically can encode if a provider supplies the fact but also gets flipped to true for confirmed user reports. If we can update it to sensitive that would be ideal. It again would take more careful planning and potentially several PRs to do it in a backwards compatible way (similar to a database column rename, really).

Reviewer 2: Although, even more ideally would be to rename it to `has_confirmed_sensitivity_report` and then when we ingest sensitivity from providers in the future to use yet another field for that. If we can make the name more explicit and prevent overloading the meaning in the future, that would be 100% better than the current ambiguous name.

Author: That makes sense! It seems like it might be outside the scope of this IP though, so I'll leave this as-is for now unless there are any objections.

Reviewer 2: Totally, I did not think it was something we would handle in this. Totally fine by me to continue excluding it from the scope.

#### Copy

Most changes necessary for updating the copy in the Django Admin UI will be
reflected by the code changes that are made above (since the Admin UI
introspects the code directly). The only other references that will need to be

> updated are the `help_text` values in the following places:

Author: Right, but I think they will be migrations without SQL?

> updated are the `help_text` values in the following places:

Reviewer 2: Yup, they'll be noop migrations. Same with changing the choices list for a field.

- [Search store references](https://github.com/WordPress/openverse/blob/92ea64c6d4ac1801785eadf99a3782628eadad06/frontend/src/stores/search.ts)
- [Content report utilities](https://github.com/WordPress/openverse/blob/74da3d852a45c514c0159038c86164a26d70504b/frontend/src/utils/content-safety.ts#L10-L23)
- All affected tests.

The following references _will not be_ affected because they correspond to
either the data model or to the API query parameter `mature`. The query
parameter will be changed to `include_sensitive_results` in the API
[as part of another project](https://docs.openverse.org/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.html#filtering)

- [Search query transform utilities](https://github.com/WordPress/openverse/blob/92ea64c6d4ac1801785eadf99a3782628eadad06/frontend/src/utils/search-query-transform.ts#L31)
  _(query parameter)_
- [`Media` interface attributes](https://github.com/WordPress/openverse/blob/930a2ec4583d53e2f17e7aa397dc9b3b32cc8ebb/frontend/src/types/media.ts#L48)
  _(data model)_

#### Copy

Almost all copy changes necessary can be made in the

> [`en.json5` base translations file](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/frontend/src/locales/scripts/en.json5#L1-L0).

Reviewer 3: ```suggestion
[`en.json5` base translations file](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/frontend/src/locales/scripts/en.json5#L487-L491).
```

Author: Ah, I meant to link the whole file here, will fix!

   models in place for the time being. Data should now only be written to and
   read from the new tables.
7. Deploy this version of the API so that the only versions currently deployed
   are writing to the new tables.
8. Remove the old models (`MatureImage`, `MatureAudio`, `NsfwReport`,
   `NsfwReportAudio`) from the API codebase and deploy this version of the API.
   This will remove the old tables from the database.

Per the
[detecting sensitive textual content project plan](https://docs.openverse.org/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.html#designation-of-results-with-sensitive-terms),
the `mature` field of the media serializer will remain, and will eventually have
a `sensitivity` field added to it to encapsulate specific information about the
reason behind a media being marked sensitive. Those changes are not under the
purview of this IP.

The Elasticsearch model property
[`mature`](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/ingestion_server/ingestion_server/elasticsearch_models.py#L118)

> will also not be affected as part of this effort.

Author: Oh that is an extremely good call and great find, I'll be sure to add this!

This work will include several migrations for the API, which will need to be
deployed one after another in production as described in
[the API section](#api).

### Tools & packages

<!-- Describe any tools or packages which this work might be dependent on. If multiple options are available, try to list as many as are reasonable with your own recommendation. -->

No other tools or packages should be required.

### Other projects or work

<!-- Note any projects this plan is dependent on. -->

This work intersects with, but is distinct from, the
["Detecting and Blurring Sensitive Textual Content" project](https://docs.openverse.org/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/index.html).

## Design

<!-- Note any design requirements for this plan. -->

No design work should be necessary.

## Parallelizable streams

<!-- What, if any, work within this plan can be parallelized? -->

The frontend copy work and the API copy & code work can be done simultaneously.
The frontend code work should wait until the API work is complete so the content
reporting using `sensitive` rather than `mature` is available on the API prior

> to the frontend code changes.

Reviewer 2: This probably needs to take into account the changes that will happen in #2118, primarily to note that we'll need to coordinate around that.

I don't think the frontend copy changes need to be blocked on the API work though, which seems to be what the first sentence in this section says.

Not a blocker, just clarification. I think this coordination will happen naturally as the work on the frontend copy gets done. Maybe just adding a note to the frontend issues to be aware that the other project will also be making related changes and to be aware of potential overlaps/merge conflicts that might arise.

# 2023-05-17 Implementation Plan: Copy updates `mature` -> `sensitive`

**Author**: @aetherunbound

<!-- See the implementation plan guide for more information: https://github.com/WordPress/openverse/tree/19791f51c063d0979112f4b9f4eeace04c8cf5ff/docs/projects#implementation-plans-status-in-rfc -->
<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->

## Reviewers

<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->

- [ ] @krysal
- [ ] @dhruvkb

> 

Reviewer 2: ```suggestion
- [x] @sarayourfriend

```

Might as well add my approval for the record.

- [`MatureAudio`](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/api/api/models/audio.py#L263)

The
[`mature` reason](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/api/api/models/media.py#L22-L21)
(and
[`mature_filtered`](https://github.com/WordPress/openverse/blob/2041c5df1e9d5d1f9f37e7c177f2e70f61ea5dba/api/api/models/media.py#L18-L17)
value) in `AbstractMediaReport` definition will also need to be changed to
`sensitive` and `sensitive_filtered` respectively. The migration necessary for
changing the data is described below, however the API will need to continue to
accept `mature` as a report reason and convert it internally to `sensitive` (at
least until we decide to make a version change to the API). The media report
tables already use table aliases as well
([`ImageReport`](https://github.com/WordPress/openverse/blob/7b95a4c8eaa9804f53b4be7ac969e04ca437695a/api/api/models/image.py#L121-L122)
and
[`AudioReport`](https://github.com/WordPress/openverse/blob/7b95a4c8eaa9804f53b4be7ac969e04ca437695a/api/api/models/audio.py#L304-L305)) -

> since we're already performing table renames as part of this IP, it makes sense

Reviewer: ```suggestion
since we're already performing table renames as part of this implementation plan, it makes sense
```

2. Modify the API to write to both the old and new tables for each of the above
   instances. On the new report tables, the `mature_filtered` reason should be
   written as `sensitive_filtered`. Data should continue to be read from the
   original tables during this time.
3. Deploy this version of the API.
4. Create a
   [data management command](https://docs.openverse.org/general/zero_downtime_database_management.html#django-management-command-based-data-transformations)
   which copies all data from the original tables to the new tables. The
   `identifier` field can be used for determining which rows have already been
   copied.
5. Deploy this version of the API and run the data management command until no
   more rows need to be copied for each of the columns.
6. Modify the ingestion server to use the new tables as reference for
   determining mature status. Specifically the
   [`exists_in_mature` table check](https://github.com/WordPress/openverse/blob/765a24028baa922b27fc0c38b8ae4c69902eec40/ingestion_server/ingestion_server/queries.py#L33-L37)
   (this will necessarily involve updating tests and initialization SQL as
   well).

> 7. Remove all code references to the old model from the API, but leave the

Reviewer: Which model exactly are we speaking of here? Is it in the code from the Ingestion server?

Author: This refers to the `MatureImage` and `MatureAudio` models, which are replaced by `SensitiveImage` and SensitiveAudio`. I'll make that explicit here. All the steps here deal with the API, unless explicitly stated like in step 6.

   models in place for the time being. Data should now only be written to and
   read from the new tables.
8. Deploy this version of the API so that the only versions currently deployed
   are writing to the new tables.
9. Remove the old models (`MatureImage`, `MatureAudio`, `NsfwReport`,
   `NsfwReportAudio`) from the API codebase and deploy this version of the API.
   This will remove the old tables from the database.

Per the
[detecting sensitive textual content project plan](https://docs.openverse.org/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.html#designation-of-results-with-sensitive-terms),
the `mature` field of the media serializer will remain, and will eventually have
a `sensitivity` field added to it to encapsulate specific information about the
reason behind a media being marked sensitive. Those changes are not under the

> purview of this IP.

Reviewer: ```suggestion
purview of this implementation plan.
```
